data/transactions/logic/assembler.go

```diff
@@ -1184,17 +1150,177 @@ func typeLoads(pgm *ProgramKnowledge, args []string) (StackTypes, StackTypes) {
 	// parse basics.Address, actually just another []byte constant
-	"addr": {0, "addr", nil, proto(":b"), 1, assembler(asmAddr)},
+	"addr": {anyImmediates: OpSpec{Name: "addr", Version: 1, Proto: proto(":b"), OpDetails: assembler(asmAddr)}},
 	// take a signature, hash it, and take first 4 bytes, actually just another []byte constant
-	"method": {0, "method", nil, proto(":b"), 1, assembler(asmMethod)},
+	"method":  {anyImmediates: OpSpec{Name: "method", Version: 1, Proto: proto(":b"), OpDetails: assembler(asmMethod)}},
+	"txn":     {1: OpSpec{Name: "txn"}, 2: OpSpec{Name: "txna"}},
+	"gtxn":    {2: OpSpec{Name: "gtxn"}, 3: OpSpec{Name: "gtxna"}},
+	"gtxns":   {1: OpSpec{Name: "gtxns"}, 2: OpSpec{Name: "gtxnsa"}},
+	"extract": {0: OpSpec{Name: "extract3"}, 2: OpSpec{Name: "extract"}},
+	"replace": {0: OpSpec{Name: "replace3"}, 1: OpSpec{Name: "replace2"}},
+}
+
+func addPseudoDocTags() {
+	for name, specs := range pseudoOps {
+		for i, spec := range specs {
+			if spec.Name == name {
+				continue
+			}
+			msg := ""
+			switch {
+			case i > 1:
+				msg = fmt.Sprintf("`%s` can be called using `%s` with %d immediates.", spec.Name, name, i)
+			case i == 1:
+				msg = fmt.Sprintf("`%s` can be called using `%s` with %d immediate.", spec.Name, name, i)
+			case i == 0:
+				msg = fmt.Sprintf("`%s` can be called using `%s` without immediates.", spec.Name, name)
+			default:
+				continue
+			}
+			desc, ok := opDocByName[spec.Name]
+			if ok {
+				opDocByName[spec.Name] = desc + "<br />" + msg
+			} else {
+				opDocByName[spec.Name] = msg
+			}
+		}
+	}
+}
+
+func init() {
+	addPseudoDocTags()
+}
+
+// mergeProtos allows us to support typetracking of pseudo-ops which are given an improper number of immediates
+//by creating a new proto that is a combination of all the pseudo-op's possibilities
+func mergeProtos(specs map[int]OpSpec) (Proto, uint64, bool) {
+	var args StackTypes
+	var returns StackTypes
+	var minVersion uint64
+	i := 0
+	for _, spec := range specs {
+		if i == 0 {
+			args = spec.Arg.Types
+			returns = spec.Return.Types
+			minVersion = spec.Version
+		} else {
+			if spec.Version < minVersion {
+				minVersion = spec.Version
+			}
+			if len(args) != len(spec.Arg.Types) || len(returns) != len(spec.Return.Types) {
+				return Proto{}, 0, false
+			}
+			for j := range args {
+				if args[j] != spec.Arg.Types[j] {
+					args[j] = StackAny
+				}
+			}
+			for j := range returns {
+				if returns[j] != spec.Return.Types[j] {
+					returns[j] = StackAny
+				}
+			}
+		}
+		i++
+	}
+	return Proto{typedList{args, ""}, typedList{returns, ""}}, minVersion, true
+}
+
+func prepareVersionedPseudoTable(version uint64) map[string]map[int]OpSpec {
+	m := make(map[string]map[int]OpSpec)
+	for name, specs := range pseudoOps {
+		m[name] = make(map[int]OpSpec)
+		for numImmediates, spec := range specs {
+			if spec.Version != 0 {
+				// If a version is specified, then we assume it is a custom spec
```

Comment:
@iten-alg Leaving a note to document our verbal with @jannotti:  We discussed using `Proto` rather than `Version` to mark a custom specification.

Reply:
@iten-alg Resolving - Thanks for bringing in the feedback via https://github.com/algorand/go-algorand/pull/4170/commits/049f614c69b8d0e11f2f9a21ce7d95441c9cc320.